File: 20.ValidParenthesis/Solution.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
public:
    inline bool ismatch(char c, char d)
    {
        if ((c == ')' and d == '(') or (c == ']' and d == '[') or (c == '}' and d == '{'))
            return true;
        return false;
    }
    bool isValid(string s)
    {
        stack<char> stk;
        for (char c : s)
        {
            if (c == '(' or c == '[' or c == '{')
                stk.push(c);
            else if (stk.empty() || !ismatch(c, stk.top()))
                return false;
            else
                stk.pop();
        }
        return stk.empty();
    }
};

File: 20.ValidParenthesis/Solution.py
class Solution:
    def isMatch(c, d):
        if ((c == ')' and d == '(') or (c == ']' and d == '[') or (c == '}' and d == '{')):
            return True
        return False

    def isValid(self, s: str) -> bool:
        stk = []
        for c in s:
            if c == '(' or c == '[' or c == '{':
                stk.append(c)
            elif len(stk) == 0 or not self.isMatch(c, stk[-1]):
                return False
            else:
                stk.pop()

        return len(stk) is 0
